server/auth/types/authentication_type.ts

```diff
@@ -38,8 +46,176 @@ export type IAuthHandlerConstructor = new (
+    //   1. check if request includes auth header or paramter(e.g. jwt in url params) is present, if so, authenticate with auth header.
+    //   2. if auth header not present, check if auth cookie is present, if no cookie, send to authentiation workflow
+    //   3. verify whether auth cookie is valid, if not valid, send to authentication workflow
+    //   4. if cookie is valid, pass to route handlers
+    const authHeaders = {};
+    let cookie: SecuritySessionCookie | null | undefined;
+    let authInfo: any | undefined;
+    // if this is an REST API call, suppose the request includes necessary auth header
+    // see https://www.elastic.co/guide/en/kibana/master/using-api.html
+    if (this.isRestApiCall(request) || this.requestIncludesAuthInfo(request)) {
+      try {
+        const additonalAuthHeader = this.getAdditionalAuthHeader(request);
+        authInfo = await this.securityClient.authinfo(request, additonalAuthHeader);
+        cookie = this.getCookie(request, authInfo);
+        this.sessionStorageFactory.asScoped(request).set(cookie);
+      } catch (error) {
+        return response.unauthorized({
+          body: error.message,
+        });
+      }
+    } else {
+      // no auth header in request, try cookie
+      try {
+        cookie = await this.sessionStorageFactory.asScoped(request).get();
+      } catch (error) {
+        this.logger.error(`Error parsing cookie: ${error.message}`);
+        cookie = undefined;
+      }
+
+      if (!cookie || !this.isValidCookie(cookie)) {
+        // clear cookie
+        this.sessionStorageFactory.asScoped(request).clear();
+        // send to auth workflow
+        return this.redirectToAuth(request, response, toolkit);
+      }
+
+      // extend cookie expiration time
+      cookie!.expiryTime = Date.now() + this.config.cookie.ttl;
+      this.sessionStorageFactory.asScoped(request).set(cookie!);
+
+      // cookie is valid
+      // build auth header
+      const authHeadersFromCookie = this.buildAuthHeaderFromCookie(cookie!);
+      Object.assign(authHeaders, authHeadersFromCookie);
+      const additonalAuthHeader = this.getAdditionalAuthHeader(request);
+      Object.assign(authHeaders, additonalAuthHeader);
+    }
+
+    // resolve tenant if necessary
+    if (this.config.multitenancy?.enabled && isMultitenantPath(request)) {
+      try {
+        const tenant = await this.resolveTenant(request, cookie!, authHeaders, authInfo);
+        // return 401 if no tenant available
+        if (!tenant) {
+          return response.badRequest({
+            body:
+              'No availabe tenant for current user, please reachout to your system administrator',
+          });
+        }
+        // set tenant in header
+        Object.assign(authHeaders, { securitytenant: tenant });
+
+        // set tenant to cookie
+        if (tenant !== cookie!.tenant) {
+          cookie!.tenant = tenant;
+          this.sessionStorageFactory.asScoped(request).set(cookie!);
```

Comment:
Is "!" after cookie needed?

Reply:
yes, it is required by typescript compiler